odl/trafos/convolution.py

```diff
@@ -0,0 +1,713 @@
+            from the kernel.
+            Note: custom range is currently not supported.
+
+        impl : string, optional
+            Implementation of the convolution. Available options are:
+
+            'scipy_convolve': Real-space convolution using
+            `scipy.signal.convolve` (default, fast for short kernels)
+
+        scale : `bool`, optional
+            If `True`, scale the discrete convolution with
+            `DiscreteLp.cell_volume`, such that it
+            corresponds to a continuous convolution.
+            Default: `True`
+
+        resample : string or sequence of strings
+            If ``kernel`` and ``domain`` have different cell sizes,
+            this option defines the behavior of the evaluation. The
+            following values can be given (per axis in the case of a
+            sequence):
+
+            'down' : Use the larger one of the two cell sizes (default).
+
+            'up' : Use the smaller one of the two cell sizes. Note that
+            this can be computationally expensive.
+
+        kernel_kwargs : dict, optional
+            Keyword arguments passed to the call of the kernel function
+            if given as a callable.
+
+        See also
+        --------
+        scipy.signal.convolve : real-space convolution
+
+        Notes
+        -----
+        - The continuous convolution of two functions
+          :math:`f` and :math:`g` defined on :math:`\mathbb{R^d}` is
+
+              :math:`[f \\ast g](x) =
+              \int_{\mathbb{R^d}} f(x - y) g(y) \mathrm{d}y`.
+
+          This implementation covers the case of a fixed convolution
+          kernel :math:`k`, i.e. the linear operator :math:`\mathcal{C}_k`
+          given by
+
+              :math:`\mathcal{C}_k(f): x \mapsto [k \\ast f](x)`.
+
+        - If the convolution kernel :math:`k` is defined on a rectangular
+          domain :math:`\Omega_0` with midpoint :math:`m_0`, then the
+          convolution :math:`k \\ast f` of a function
+          :math:`f:\Omega \\to \mathbb{R}` with :math:`k` has a support
+          that is a superset of :math:`\Omega + m_0`. Therefore, we
+          choose to keep the size of the domain and take
+          :math:`\Omega + m_0` as the domain of definition of functions
+          in the range of the convolution operator.
+
+          In fact, the support of the convolution is contained in the
+          `Minkowski sum
+          <https://en.wikipedia.org/wiki/Minkowski_addition>`_
+          :math:`\Omega + \Omega_0`.
+
+          For example, if :math:`\Omega = [0, 5]` and
+          :math:`\Omega_0 = [1, 3]`, i.e. :math:`m_0 = 2`, then
+          :math:`\Omega + \Omega_0 = [1, 8]`. However, we choose the
+          shifted interval :math:`\Omega + m_0 = [2, 7]` for the range
+          of the convolution since it contains the "main mass" of the
+          result.
+        """
+        # Basic checks
+        if not isinstance(domain, DiscreteLp):
```

Comment:
Question here: How about convolutions on `FnBase`? Would make sense mathematically, but feels like lots of `if...else` stuff that would make the code unreadable.